### PR TITLE
Improve Parallelism Defaults

### PR DESCRIPTION
**Pull request by rebeccafair**
Addresses #128

### CODE REVIEW

File: euphonic/force_constants.py [hunk 1]
```diff
@@ -445,19 +449,28 @@ def _calculate_phonons_at_qpts(
...
-                        'have been installed. Attempting to fall back '
-                        'to pure Python calculation'), stacklevel=2)
+                    if use_c is None:
+                        warnings.warn((
+                            cext_err_msg
+                            + ' Falling back to pure Python calculation.'),
+                            stacklevel=3)
                     raise
             else:
                 raise ImportError
```
Comment (ajjackson) [inline euphonic/force_constants.py:468]: I do find this stack of if functions and try/Except a bit tricky to follow. I'm also wary also of using try/except/raise as a form of control logic over a large block of code. What happens if something other than `import euphonic._euphonic` raises an ImportError? Then it shouldn't be excepted. I think this was motivated by the problem of reaching the python implementation from different paths. How about instead we put that in a function and have something like:

```python
if use_c is not False:
    try:
        import euphonic._euphonic as euphonic_c
    except ImportError:
        if use_c:
            raise
        else:
            warnings.warn(...)
            _do_it_with_python(*args)
    else:
        _do_it_with_c(*args)
else:
    _do_it_with_python(*args)
```

**Review by ajjackson — CHANGES_REQUESTED** [review]
I agree with the approach here. The new tests look great (and taught me something neat about testing warnings!)

The threadctl context manager is very cool, and a non-intrusive way of dealing with this problem.

I just raised one concern about the try/except logic which is a bit hard to follow and also could be safer.

File: euphonic/force_constants.py (hunk 1 above)
Comment (rebeccafair) [reply euphonic/force_constants.py]: Ok done! I've implemented it a little differently, as I found handling the different output arguments for Python implementation (depending on whether `return_eigenvectors=True/False`) was kind of awkward. But I think it works!

File: euphonic/force_constants.py [hunk 2]
```diff
@@ -453,19 +453,23 @@ def _calculate_phonons_at_qpts(
...
+                        cext_err_msg
+                        + ' Falling back to pure Python calculation.'),
+                        stacklevel=3)
+                else:
+                    raise ImportCError(cext_err_msg)
+
+        if use_c_status is True:
             if n_threads is None:
                 try:
                     n_threads = int(os.environ['EUPHONIC_NUM_THREADS'])
```
Comment (ajjackson) [inline euphonic/force_constants.py:475]: I think this will raise a ValueError if the user has set `EUPHONIC_NUM_THREADS=''`; setting to '' is a common way of pseudo-clearing an environment variable. (It will return the same thing to bash as if the variable was unset.) It will also ValueError on other invalid values (e.g. `'TWO'`) but in that case raising the error is probably best.

As well as accepting `''`, we can avoid the try/except for KeyError with `os.environ.get()`:
```python
n_threads_env = os.environ.get('EUPHONIC_NUM_THREADS', '')   # type: str
if n_threads_env:
    n_threads = int(n_threads_env)
else:
    n_threads = cpu_count()
```

File: euphonic/force_constants.py [hunk 3]
```diff
@@ -453,19 +453,23 @@ def _calculate_phonons_at_qpts(
...
-                        warnings.warn((
-                            cext_err_msg
-                            + ' Falling back to pure Python calculation.'),
-                            stacklevel=3)
-                    raise
-            else:
-                raise ImportError
+
+        # Check if C extension can be used and handle appropriately
+        use_c_status = False
```
Comment (ajjackson) [inline euphonic/force_constants.py:458]: Good call. Using a flag like this isn't as "pretty" as the other approaches we considered, but in this case it's clearly the most readable.

**Review by ajjackson — APPROVED** [review]
Looks good! I flagged up one last tweak, sorry :sweat:

File: euphonic/force_constants.py (hunk 2 above)
Comment (rebeccafair) [reply euphonic/force_constants.py]: Good call, I didn't think of this case
